gcc/rust/lex/rust-lex.cc

```diff
@@ -2504,338 +2499,133 @@ Lexer::parse_char_or_lifetime (Location loc)
-	      if ((input3 & 0xC0) != 0x80)
-		  return 0xFFFE;
-
-	      uint32_t output
-		= ((input & 0x0F) << 12) | ((input2 & 0x3F) << 6) | ((input3 &
-     0x3F) << 0); return output; } else if ((input & 0xF8) == 0xF0) {
-	      // 4 bytes
-	      uint8_t input2 = peek_input(1);
-	      if ((input2 & 0xC0) != 0x80)
-		  return 0xFFFE;
-
-	      uint8_t input3 = peek_input(2);
-	      if ((input3 & 0xC0) != 0x80)
-		  return 0xFFFE;
-
-	      uint8_t input4 = peek_input(3);
-	      if ((input4 & 0xC0) != 0x80)
-		  return 0xFFFE;
-
-	      uint32_t output = ((input & 0x07) << 18) | ((input2 & 0x3F) << 12)
-				| ((input3 & 0x3F) << 6) | ((input4 & 0x3F) <<
-     0); return output; } else { rust_error_at(get_current_location(), "invalid
-     UTF-8 (too long)"); return 0xFFFE;
-	  }*/
+  Rust::Lexer::BufferInputSource source (str, 0);
+  assert_source_content (source, expected);
 }
 
 void
-Lexer::split_current_token (TokenId new_left, TokenId new_right)
+test_file_input_source (std::string str, std::vector<uint32_t> expected)
 {
-  /* TODO: assert that this TokenId is a "simple token" like punctuation and not
-   * like "IDENTIFIER"? */
-  Location current_loc = peek_token ()->get_locus ();
-  TokenPtr new_left_tok = Token::make (new_left, current_loc);
-  TokenPtr new_right_tok = Token::make (new_right, current_loc + 1);
-
-  token_queue.replace_current_value (std::move (new_left_tok));
-  token_queue.insert (1, std::move (new_right_tok));
+  FILE *tmpf = tmpfile ();
+  // Moves to the first character
+  fputs (str.c_str (), tmpf);
+  std::rewind (tmpf);
+  Rust::Lexer::FileInputSource source (tmpf);
+  assert_source_content (source, expected);
 }
 
 void
-Lexer::start_line (int current_line, int current_column)
+rust_input_source_test ()
 {
-  if (line_map)
-    line_map->start_line (current_line, current_column);
+  // ASCII
+  std::string src = u8"_abcde\tXYZ\v\f";
+  std::vector<uint32_t> expected
+    = {'_', 'a', 'b', 'c', 'd', 'e', '\t', 'X', 'Y', 'Z', '\v', '\f'};
+  test_buffer_input_source (src, expected);
+
+  // BOM
+  src = u8"\xef\xbb\xbfOK";
+  expected = {'O', 'K'};
+  test_buffer_input_source (src, expected);
+
+  // Russian
+  src = u8"приве́т";
+  expected = {L'п',
+	      L'р',
+	      L'и',
+	      L'в',
+	      0x0435 /* CYRILLIC SMALL LETTER IE е */,
+	      0x301 /* COMBINING ACUTE ACCENT ́ */,
+	      L'т'};
+  test_buffer_input_source (src, expected);
+
+  src = u8"❤️🦀";
+  expected = {0x2764 /* HEAVY BLACK HEART */,
+	      0xfe0f /* VARIATION SELECTOR-16 */, L'🦀'};
+  test_buffer_input_source (src, expected);
+
+  src = u8"こんにちは";
+  expected = {L'こ', L'ん', L'に', L'ち', L'は'};
+  test_file_input_source (src, expected);
+
+  src = u8"👮‍♂👩‍⚕";
+  expected
+    = {0x1f46e /* POLICE OFFICER */,   0x200d /* ZERO WIDTH JOINER */,
+       0x2642 /* MALE SIGN */,	       0x1f469 /* WOMAN */,
+       0x200d /* ZERO WIDTH JOINER */, 0x2695 /* STAFF OF AESCULAPIUS */};
+  test_file_input_source (src, expected);
```

Comment:
This is one of the tests for `BufferInputSource`, which internally calls `std::tmpfile()`.